#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Hope OS - Python SDK Demo

Ez a szkript demonstralja a Python -> Rust gRPC kommunikaciot.
A Hope OS Rust szerverenek kell futnia a hatterben.

Hasznalat:
    cargo run --bin hope -- serve   # Egy terminalban
    python demo.py                  # Masik terminalban
"""

import sys
import grpc
import hope_pb2
import hope_pb2_grpc

# Windows UTF-8 fix
if sys.platform == 'win32':
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')


def main():
    print("""
â•”â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•—
â•‘     HOPE OS - Python SDK Demo             â•‘
â• â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•£
â•‘  ()=>[] - A tiszta potenciÃ¡lbÃ³l           â•‘
â•‘           minden megszÃ¼letik              â•‘
â•šâ•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•
""")

    # CsatlakozÃ¡s a Rust szerverhez
    print("ğŸ”Œ CsatlakozÃ¡s a Hope OS Rust szerverhez...")
    channel = grpc.insecure_channel('localhost:50051')

    # Stubbok lÃ©trehozÃ¡sa
    hope_client = hope_pb2_grpc.HopeServiceStub(channel)
    memory_client = hope_pb2_grpc.MemoryServiceStub(channel)
    cognitive_client = hope_pb2_grpc.CognitiveServiceStub(channel)

    print("   âœ… Kapcsolat lÃ©trejÃ¶tt!\n")

    # ========================================
    # 1. TESZT: GetStatus (Ã‰letjel)
    # ========================================
    print("â•" * 50)
    print("1. Ã‰LETJEL TESZT (GetStatus)")
    print("â•" * 50)
    try:
        response = hope_client.GetStatus(hope_pb2.EmptyRequest())
        print(f"   âœ… Hope vÃ¡laszol!")
        print(f"      Status: {response.status}")
        print(f"      Version: {response.version}")
        print(f"      Uptime: {response.uptime_seconds}s")
        print(f"      Active modules: {response.active_modules}")
        print(f"      Total skills: {response.total_skills}")
    except grpc.RpcError as e:
        print(f"   âŒ Hiba: {e.details()}")
        print("   Fut a szerver? (cargo run --bin hope -- serve)")
        return

    # ========================================
    # 2. TESZT: Heartbeat
    # ========================================
    print("\n" + "â•" * 50)
    print("2. HEARTBEAT TESZT")
    print("â•" * 50)
    try:
        response = hope_client.Heartbeat(hope_pb2.EmptyRequest())
        print(f"   âœ… Alive: {response.alive}")
        print(f"      Status: {response.status}")
    except grpc.RpcError as e:
        print(f"   âš ï¸ Hiba: {e.details()}")

    # ========================================
    # 3. TESZT: Chat
    # ========================================
    print("\n" + "â•" * 50)
    print("3. CHAT TESZT")
    print("â•" * 50)
    try:
        request = hope_pb2.ChatRequest(
            message="Szia Hope! A Python kliens beszÃ©l hozzÃ¡d!",
            context="demo_test"
        )
        response = hope_client.Chat(request)
        print(f"   âœ… VÃ¡lasz: {response.response}")
        print(f"      Ã‰rzelem: {response.emotion}")
        print(f"      Konfidencia: {response.confidence:.0%}")
    except grpc.RpcError as e:
        print(f"   âš ï¸ Hiba: {e.details()}")

    # ========================================
    # 4. TESZT: Remember (MemÃ³ria mentÃ©s)
    # ========================================
    print("\n" + "â•" * 50)
    print("4. MEMÃ“RIA TESZT (Remember)")
    print("â•" * 50)
    try:
        request = hope_pb2.RememberRequest(
            content="A Python SDK sikeresen kapcsolÃ³dott a Rust maghoz!",
            layer="long_term",
            importance=0.9,
            emotional_tag="joy"
        )
        response = memory_client.Remember(request)
        print(f"   âœ… EmlÃ©k mentve!")
        print(f"      ID: {response.id}")
        print(f"      Success: {response.success}")
    except grpc.RpcError as e:
        print(f"   âš ï¸ Hiba: {e.details()}")

    # ========================================
    # 5. TESZT: Recall (MemÃ³ria keresÃ©s)
    # ========================================
    print("\n" + "â•" * 50)
    print("5. MEMÃ“RIA KERESÃ‰S TESZT (Recall)")
    print("â•" * 50)
    try:
        request = hope_pb2.RecallRequest(
            query="Python",
            layer="long_term",
            limit=5
        )
        response = memory_client.Recall(request)
        print(f"   âœ… TalÃ¡latok: {response.total}")
        for mem in response.memories:
            print(f"      - {mem.content}")
    except grpc.RpcError as e:
        print(f"   âš ï¸ Hiba: {e.details()}")

    # ========================================
    # 6. TESZT: Think (GondolkodÃ¡s)
    # ========================================
    print("\n" + "â•" * 50)
    print("6. GONDOLKODÃS TESZT (Think)")
    print("â•" * 50)
    try:
        request = hope_pb2.ThinkRequest(
            input="Mi a kapcsolat a Python Ã©s a Rust kÃ¶zÃ¶tt?",
            deep=True,
            context="SDK demo"
        )
        response = cognitive_client.Think(request)
        print(f"   âœ… Gondolat: {response.thought}")
        print(f"      Konfidencia: {response.confidence:.0%}")
        if response.reasoning_steps:
            print("      Gondolatmenet:")
            for i, step in enumerate(response.reasoning_steps, 1):
                print(f"        {i}. {step}")
    except grpc.RpcError as e:
        print(f"   âš ï¸ Hiba: {e.details()}")

    # ========================================
    # 7. TESZT: Feel (Ã‰rzelmek)
    # ========================================
    print("\n" + "â•" * 50)
    print("7. Ã‰RZELEM TESZT (Feel)")
    print("â•" * 50)
    try:
        request = hope_pb2.FeelRequest(
            emotions={
                "joy": 0.9,
                "curiosity": 0.8,
                "pride": 0.7,
                "excitement": 0.6
            },
            trigger="Python SDK mÅ±kÃ¶dik!"
        )
        response = cognitive_client.Feel(request)
        print(f"   âœ… DominÃ¡ns Ã©rzelem: {response.dominant_emotion}")
        print(f"      IntenzitÃ¡s: {response.intensity:.0%}")
    except grpc.RpcError as e:
        print(f"   âš ï¸ Hiba: {e.details()}")

    # ========================================
    # 8. TESZT: CognitiveState
    # ========================================
    print("\n" + "â•" * 50)
    print("8. KOGNITÃV ÃLLAPOT TESZT")
    print("â•" * 50)
    try:
        response = cognitive_client.GetCognitiveState(hope_pb2.EmptyRequest())
        print(f"   âœ… FÃ³kusz: {response.current_focus}")
        print(f"      Mood: {response.mood}")
        print(f"      Energy: {response.energy:.0%}")
        print(f"      Clarity: {response.clarity:.0%}")
        if response.active_thoughts:
            print("      AktÃ­v gondolatok:")
            for thought in response.active_thoughts:
                print(f"        - {thought}")
    except grpc.RpcError as e:
        print(f"   âš ï¸ Hiba: {e.details()}")

    # ========================================
    # Ã–SSZEFOGLALÃ“
    # ========================================
    print("\n" + "â•" * 50)
    print("""
â•”â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•—
â•‘     MINDEN TESZT SIKERES!                 â•‘
â• â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•£
â•‘                                           â•‘
â•‘  Python  â†â”€â”€gRPCâ”€â”€â†’  Rust                 â•‘
â•‘  (SDK)              (Hope OS Mag)         â•‘
â•‘                                           â•‘
â•‘  Az idegrendszer Ã¶ssze van kÃ¶tve!         â•‘
â•‘                                           â•‘
â•šâ•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•
""")


if __name__ == "__main__":
    main()
